File: backend/app/api/routes/opportunities.py
from fastapi import APIRouter, Depends, Query

from app.api.deps import get_runtime
from app.schemas.api import OpportunityListResponse, PriorityFilter
from app.services.opportunities import list_opportunities
from app.services.runtime import AppRuntime

router = APIRouter(prefix="/opportunities", tags=["opportunities"])


@router.get("", response_model=OpportunityListResponse)
def get_opportunities(
    priority: PriorityFilter | None = Query(default=None),
    status: str = Query(default="open"),
    minimum_amount: float | None = Query(default=None, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    runtime: AppRuntime = Depends(get_runtime),
) -> OpportunityListResponse:
    return list_opportunities(
        runtime,
        priority=priority,
        status=status,
        minimum_amount=minimum_amount,
        limit=limit,
        offset=offset,
    )

File: frontend/src/api/opportunities.ts
import { apiGet } from "./client"
import type { OpportunityListResponse, OpportunityQuery } from "./types"

export function fetchOpportunities(query: OpportunityQuery = {}): Promise<OpportunityListResponse> {
  const params = new URLSearchParams()
  if (query.priority) params.set("priority", query.priority)
  if (query.status) params.set("status", query.status)
  if (query.minimum_amount != null) params.set("minimum_amount", String(query.minimum_amount))
  if (query.limit != null) params.set("limit", String(query.limit))
  if (query.offset != null) params.set("offset", String(query.offset))
  const suffix = params.toString() ? `?${params.toString()}` : ""
  return apiGet<OpportunityListResponse>(`/api/v1/opportunities${suffix}`)
}
